### DISCUSSION TITLE
enum type can be case-insensitive-mapped with json?

### DISCUSSION THREAD

**Discussion by YunchengLiu** [nlohmann/json#3987] [Ideas]
`NLOHMANN_JSON_SERIALIZE_ENUM` provides a tool to map enum and json precisely. I want to know if it can provide a arg to specify enabling case-insensitive-mapping?

I think that most of the cases where enum is serialized to json should be ascii strings, so it should be more convenient to provide case-independent conversion in some scenarios, and there should be no need to consider UNICODE at this time.

```c++
#include <algorithm>
#include <string_view>

constexpr bool iequal(std::string_view lhs, std::string_view rhs) noexcept {
        return std::ranges::equal(lhs, rhs, [](char const c1, char const c2) noexcept{
                return std::tolower(c1) == std::tolower(c2);
            });
    }

#define NLOHMANN_JSON_CASE_INSENSITIVE_SERIALIZE_ENUM(EnumType, ...)            \
    template<typename JsonType>                                                 \
    inline void to_json(JsonType& j, EnumType const& e) {                       \
        static_assert(std::is_enum_v<EnumType>, #EnumType " must be an enum!"); \
        static std::pair<EnumType, JsonType> const m[] = __VA_ARGS__;           \
        auto it  = std::ranges::find_if(m, [e](auto const& obj) noexcept {      \
            return obj.first == e;                                              \
        });                                                                     \
        j = ((it != std::end(m)) ? it : std::begin(m))->second;                 \
    }                                                                           \
    template<typename JsonType>                                                 \
    inline void from_json(JsonType const& j, EnumType& e) {                     \
        static_assert(std::is_enum_v<EnumType>, #EnumType " must be an enum!"); \
        static std::pair<EnumType, JsonType> const m[] = __VA_ARGS__;           \
        auto it  = std::ranges::find_if(m, [&j](auto const& obj) noexcept{      \
            return iequal(obj.second.dump(), j.dump());       /*modify here*/   \
        });                                                                     \
        e = ((it != std::end(m)) ? it : std::begin(m))->first;                  \
    }

```
I implemented it in this simple way (as above), and there seems to be no problem. Is there any other complicated situation that I have not considered?

**Comment by nlohmann** [comment]
There is no case-insensitive variant of `NLOHMANN_JSON_SERIALIZE_ENUM` today; the macro does an exact string match against the provided pairs. This is a plausible, self-contained convenience macro (e.g. `NLOHMANN_JSON_SERIALIZE_ENUM_CASE_INSENSITIVE`), though it's a fairly narrow use case that most users can already solve by normalizing case in their own `to_json`/`from_json` before calling the standard macro, so it's more 'nice to have' than essential.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
